### ISSUE TITLE
pyright does not infer list of Literals, but mypy does

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#9491]
**Describe the bug**
If you have declared something that is a collection of literal values, then pyright doesn't infer that a comprehension over that collection can only contain those literal values.  Mypy does do this.

**Code or Screenshots**

```python
from typing import Literal

LitChoices = Literal["a", "b", "c"]

mydict: dict[int, LitChoices] = {1: "a", 2: "b", 3: "c"}

foo = list(v for k, v in mydict.items())
reveal_type(foo)

```
`mypy` reports:
```text
 Revealed type is "builtins.list[Union[Literal['a'], Literal['b'], Literal['c']]]"
```
`pyright` reports:
```text
information: Type of "foo" is "list[str]"
```

I would have expected `pyright` to do what `mypy` does.

**VS Code extension or command-line**

pyright command line 1.1.389
mypy 1.13.0

**Comment by erictraut** [comment]
Pyright is behaving as designed, so this isn't a bug. Mypy is inconsistent in how it handles literals. Pyright is consistent. You would expect `["a", "b", "c"]` to be inferred as `list[str]` rather than `list[Literal["a", "b", "c"]]`. Similarly, if you have a parameter `x` whose type is `Literal["a", "b", "c"]`, the inferred type of the expression `[x]` is `list[str]`. Refer to [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=inference-of-list-set-and-dict-expressions) for details.

**Comment by Dr-Irv** [comment]
I looked at your example in the docs where you say `mypy` is inconsistent, but I disagree with that.  The example is:
```python
def func(one: Literal[1]):
    reveal_type(one) # Literal[1]
    reveal_type([one]) # pyright: list[int], mypy: list[Literal[1]]

    reveal_type(1) # Literal[1]
    reveal_type([1]) # pyright: list[int], mypy: list[int]
```
In the first case, the _variable_ `one` has a declared type `Literal[1]`.  So you can infer that `[one]` is `list[Literal[1]]` .  But in the second case, there is no variable involved.  So the type of `1` in the `list` has to be wide as `list[int]`.  

I should also note that in the second case, `mypy` reveals the type of `1` to be `"Literal[1]?" .   Note the `?`

I don't see why you don't treat declarations of variables that are `Literal` the same way that you would treat a variable that was declared as being an instance of some class.

**Comment by erictraut** [comment]
The declared type of a symbol used within an expression is irrelevant. A symbol's declared type simply defines its upper bound and has no bearing on the type of an expression in which that variable appears. The locally-narrowed type of the symbol is used to evaluate the type of the expression. In this case, the expression `1` and `one` both evaluate to type `Literal[1]`. That means `[1]` and `[one]` and `[1 + 0]` and `[one + 0]` should all evaluate to the same type.

**Comment by Dr-Irv** [comment]
So here is why this is painful.  We have code that is within comprehensions, we create new collections that should be collections of the declared type.  Example:
```python
from typing import Literal, TypeAlias

Days: TypeAlias = Literal["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]


def oneday(daylist: list[Days], n: int) -> set[Days]:
    result = set(x for x in daylist)
    reveal_type(result)
    return result


def build_dict(daylist: list[Days]) -> dict[Days, str]:
    result = {k: k[0] for k in daylist}
    reveal_type(result)
    return result

```
This is an oversimplified example, but `pyright` complains about it because `result` has type `set[str]` in the first function and `dict[str, str]` in the second example.

Now imagine something like these comprehensions being used with various nesting, and you end up having to do `cast` all over the place.  Or create intermediate results with things like `result: dict[Days, str] = {k: k[0] for k in daylist}` to avoid using `cast` .

So what is your suggestion to do in this case, especially when you have a complex comprehension that `mypy` is happy with, but `pyright` then requires a `cast` ?

Note that if `Days` was a `class` called `DayClass`, then everything is OK:
```python
class DayClass:
    def __init__(self, day: Days):
        self._day = day

    def __repr__(self) -> str:
        return self._day
    
    def __str__(self) -> str:
        return self._day
    

def oneday_c(daylist: list[DayClass], n: int) -> set[DayClass]:
    result = set(x for x in daylist)
    reveal_type(result)
    return result


def build_dict_c(daylist: list[DayClass]) -> dict[DayClass, str]:
    result = {k: k[0] for k in daylist}
    reveal_type(result)
    return result
```

IMHO, you should treat a `Literal` declaration (or at least a `TypeAlias` that corresponds to a `Literal`) like a class.  This would make it easy to switch from a type that was a `Literal` to a type that is implemented as a class.

**Comment by erictraut** [comment]
The inference rules used by each type checker result in different tradeoffs. I can show you many examples where mypy's inference rules require additional type declarations as compared to pyright's. In my experience, pyright's inference rules require many fewer type declarations than mypy, but there will be counterexamples like the one you've highlighted above.

Pyright's rules are not going to change in this case because it would be inconsistent and would have many downstream negative consequences. If you want to use pyright, you'll need to adapt to its inference rules. In particular, if you want to constrain a variable to a type that includes a literal type argument, you'll need to declare its type as such.

```python
    result: set[Days] = set(x for x in daylist)
```

**Comment by Dr-Irv** [comment]
But that means if you had this:
```python
from typing import Literal, TypeAlias

Days: TypeAlias = Literal["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]

def myfunc(y: Days):
    print(y)

def iterate_this(daylist: list[Days], adict: dict[Days, Days]):
    for x in set([adict[x] for x in daylist]):
        myfunc(x)
```
Now in `iterate_this()`, `pyright` reports an incorrect function call for `myfunc()`.  But if using `pyright`, you have to now do something like:
```python
def iterate_this2(daylist: list[Days], adict: dict[Days, Days]):
    theset: set[Days] = set([adict[x] for x in daylist])
    for x in theset:
        myfunc(x)
```
I think the code in `iterate_this()` is easier to read (and more natural to write) than `iterate_this2()`.  

Maybe you'd consider creating an option in `pyright` to treat a `TypeAlias` that is a `Literal`  as if the declared `TypeAlias` was referring to a `class`.  In other words, you don't do any inference on operators on objects applied to an object that has a type that is a `TypeAlias` consisting of a `Literal`.  But maybe that's too complex.
